feat: archive and restore wishes

cj13579 opened the pull request:
This change adds support for archiving (and restoring) wishes into a "hidden" list. This can be super useful for tidying up a wish list for items that you have received, or items that aren't a priority for you right now, but might be something that you want to be able to reference in the future so don't want to completely remove.

When in table mode (the default) and viewing your own wish list a new "Archive" option is available:

![image](https://github.com/user-attachments/assets/c99a4a24-bbe7-413c-abe8-dc8957545ac5)

When clicked, a "Successfully archived item" messages flashes up and the item is removed from the wishlist. In a new, collapsed section titled "Archived items" all of the items archived are now visible:

![image](https://github.com/user-attachments/assets/3f19d2ab-f075-40ef-a9e5-d0f6e102a904)

Clicking the restore icon will remove the item from the archived list and make it visible again on the main wishlist. Clicking delete, will prompt you to make sure, and then delete the item like before (this prompt is also added to the delete action from the main wishlist):

![image](https://github.com/user-attachments/assets/1dc9a37c-228b-48b2-87f0-7440c152fa9d)

The same functionality exists when using card mode:

* Archiving from the wish list:
    ![image](https://github.com/user-attachments/assets/32350e2d-f797-4773-9045-89f3c956fd22)
* Delete and restore from Archive list
    ![image](https://github.com/user-attachments/assets/c4e376dc-69d5-4a62-946f-ea1fa0579bf2)

Also updated is the Wishlists page counts and details. The `totals()` function is expanded to report active and archived items. The counts on the progress bar, and the details list below only show progress towards "active" items (those not archived):

![image](https://github.com/user-attachments/assets/979ff90f-b971-42fd-a43c-84767cbd2871)

An option `AUTO_ARCHIVE_PEDGES` is also added. This might help with #170 but don't expect it to be widely used. When an item is pledged for someone, it will be automatically archived.

In my opinion, releasing this closes #164.

Wingysam commented on package.json:
```diff
@@ -7,7 +7,7 @@
     "test": "echo \"Error: no test specified\" && exit 1",
     "start": "node built/manager.js",
-    "dev": "nodemon --watch src -e ts,js --exec 'npm run build && node built/index.js'",
+    "dev": "nodemon --watch src -e ts,js,pug --exec 'npm run build && node built/index.js'",
```
Shouldn't need to restart on changes to pug files since they get read from disk on every request

cj13579 replied:
Yep, you're right. Patched back with https://github.com/Wingysam/Christmas-Community/pull/183/commits/f944297acc02290877e41bc64562be45ab64e998

Wingysam reviewed (commented):
I have some UX concerns with this. I think as a user I would want separate archives for each year rather than piling them all together. I'm also concerned with how this will interact with multiple wishlists in the future, like I might want a feature to make a private "Sam's Wishlist (2025)" and move all of the items from "Sam's Wishlist" to it.

I'm also concerned about the table of buttons being overwhelming. Maybe we could hide some of them inside a meatball/kebab menu? I'm thinking the edit, delete, move top, move bottom, and archive buttons could all be hidden behind a button.

cj13579 commented:
> I have some UX concerns with this. I think as a user I would want separate archives for each year rather than piling them all together. I'm also concerned with how this will interact with multiple wishlists in the future, like I might want a feature to make a private "Sam's Wishlist (2025)" and move all of the items from "Sam's Wishlist" to it.

@Wingysam, I don't think anything done here makes the migration to multiple lists any harder. In fact, I can see the Wishlist page actually being the same even when multiple wish lists are supported. A user object just has _n_ lists and a list would have _n_ items (of which those items are either _active_ or _archived_. The routing to open a list can just be updated to load a specific list from the Wishlists page, but still load the page that we have today (and that's extended here).

> I'm also concerned about the table of buttons being overwhelming. Maybe we could hide some of them inside a meatball/kebab menu? I'm thinking the edit, delete, move top, move bottom, and archive buttons could all be hidden behind a button.

I think that's also fair. I think that could be resolved as part of #154.

Wingysam commented on src/views/wishlist.pug:
```diff
@@ -288,6 +320,140 @@ block content
+  br
+  details
+    summary= lang('WISHLIST_ARCHIVED_ITEMS')
+    if global._CC.config.wishlist.table
```
I think this could be turned into a mixin with a couple of ifs for archive vs unarchive rather than duplicated. Something like `+wishlist(groupType, groupItems)`

cj13579 replied:
Yes, that's a very good idea. I've had a go at implementing this in https://github.com/Wingysam/Christmas-Community/pull/183/commits/b088c7b77b180e580780e2bd054ff996e05964da

Wingysam commented on src/views/wishlist.pug:
```diff
@@ -238,10 +265,10 @@ block content
                 .field.inline
                   .control.inline
-                    input.inline.button.is-warning(type='submit' value=lang('WISHLIST_DELETE'))
+                    input.inline.button.is-danger(type='submit' value=lang('WISHLIST_DELETE'))
```
Good change, thanks

Wingysam commented on src/views/wishlists.pug:
```diff
@@ -8,7 +8,7 @@ mixin wishlistDetails(wishlist, onlyShowAddedByUser)
     ul.wishlist-dropdown
       for item in wishlist
-        if onlyShowAddedByUser === undefined || item.addedBy === onlyShowAddedByUser
+        if !item.archived && onlyShowAddedByUser === undefined || item.addedBy === onlyShowAddedByUser
```
Should add some parentheses here to make it clear which way the operators go, so either
```js
(!item.archived && onlyShowAddedByUser === undefined) || item.addedBy === onlyShowAddedByUser
```
or
```js
!item.archived && (onlyShowAddedByUser === undefined || item.addedBy === onlyShowAddedByUser)
```

cj13579 replied:
Yep. Updated with https://github.com/Wingysam/Christmas-Community/pull/183/commits/eae7599d23c06e513dad0635347b54cc0329e898

Wingysam reviewed (commented):
Looking good so far!

I'd like to have the stashing of buttons in a menu merged before this to avoid making the table look too complicated in the mean time.

cj13579 commented:
@Wingysam what about something like this?

![image](https://github.com/user-attachments/assets/957bce92-60cb-4fee-97f1-b0c0c3f56b11)

Top row is what it looks like when you've clicked on the hamburger, second row is what it looks like "normally". Hovering over each icon gives to the localized text of what the icon represents (though I think they do a good job at representing what they mean). The only thing that I have noticed with my change is that the up/down actions now error. Top and bottom work, but the individual moves don't.

You're pretty handy with that stuff, would you mind taking a look?


Also, mobile view:

- Collapsed:
    ![image](https://github.com/user-attachments/assets/29879205-85fa-4d8b-a6b2-aa782e1014d5)
- Expanded:
    ![image](https://github.com/user-attachments/assets/0781e94a-97bc-4cd9-8810-a10415344a7a)

Wingysam commented:
I think that's moving in the right direction! My vision for it is something similar to the Discord menu like this:
<img width="262" alt="image" src="https://github.com/user-attachments/assets/87e3071c-21da-46f9-a2cd-7e1bf0b92a1b" />

I think Bulma has some built-in components for that. I should be able to try it at some point.

cj13579 commented:
@Wingysam better?

![image](https://github.com/user-attachments/assets/d45990f3-1eb8-40f5-9d74-b6b4dba94262)

cj13579 commented:
@Wingysam in addition to the discord looking buttons, I have added a new "super admin" feature. I think its a fair discussion as to whether this happens outside of this PR, but I've added it here for now anyway - here's what it does:

When the environment variable "SUPER_ADMINS" is set to `true`, admins can:

1. See _all_ items on a Wishlist, including those added by others to their own list
2. Manage other people's lists including update rankings, edit items, and archive items

The use case for this for me is that I manage a list for my sons. The list is a year-round thing (birthday, Christmas, etc.). I'd like to archive off the items that they received this past Christmas, and tidy everything up before myself and others add new things for their upcoming birthdays.

The default behavior of the application remains the same (admins are not "super").

cj13579 commented on src/routes/wishlist/index.js:
```diff
@@ -63,8 +63,15 @@ export default function (db) {
+        var items = null
         await wishlist.fetch()
-        const items = await wishlist.itemsVisibleToUser(req.user._id)
+        if (req.user.admin && _CC.config.superAdmins) {
```
This should be updated to pass this logic into the `itemsVisibleToUser` function. I'll do that.
